docs/_ext/github_links.py

```diff
@@ -0,0 +1,136 @@
+import ast
+import functools
+import importlib.util
+import pathlib
+
+BASE_PATH = pathlib.Path(__file__).parents[2]
+
+
+class CodeLocator(ast.NodeVisitor):
+    def __init__(self):
+        self.current_path = []
+        self.node_line_numbers = {}
+        self.import_locations = {}
+
+    @classmethod
+    def from_code(cls, code):
+        tree = ast.parse(code)
+        locator = cls()
+        locator.visit(tree)
+        return locator
+
+    def visit_node(self, node):
+        self.current_path.append(node.name)
+        self.node_line_numbers[".".join(self.current_path)] = node.lineno
+        ast.NodeVisitor.generic_visit(self, node)
+        self.current_path.pop()
+
+    def visit_FunctionDef(self, node):
+        self.visit_node(node)
+
+    def visit_ClassDef(self, node):
+        self.visit_node(node)
+
+    def visit_ImportFrom(self, node):
+        for alias in node.names:
+            if alias.asname:
+                # Exclude linking aliases (`import x as y`) to avoid confusion when
+                # clicking a source link to a differently named entity.
+                continue
+            self.import_locations[alias.name] = ("." * node.level) + (node.module or "")
+
+
+@functools.lru_cache(maxsize=1024)
+def get_locator(file: pathlib.Path) -> CodeLocator:
+    file_contents = file.read_text()
```

Comment:
```suggestion
    file_contents = file.read_text(encoding="utf-8")
```
@smithdc1 is this also an option?

Reply:
Yup, that'll do it 👍